### PR TITLE
[cpp] support rt-thread CPP wrapper sub-switch macros

### PR DESCRIPTION
**Pull request by yangjie11**
## 拉取/合并请求描述：(PR description)

[

#### 在什么测试环境下测试通过 (what is the test environment)

]

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [ ] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [ ] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [ ] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [ ] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other styles
- [ ] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [ ] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [ ] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [ ] 代码是高质量的 Code in this PR is of high quality
- [ ] 已经使用[formatting](https://github.com/mysterywolf/formatting)等源码格式化工具确保格式符合[RT-Thread代码规范](https://github.com/RT-Thread/rt-thread/blob/master/documentation/contribution_guide/coding_style_cn.md) This PR complies with [RT-Thread code specification](https://github.com/RT-Thread/rt-thread/blob/master/documentation/contribution_guide/coding_style_en.md)

### CODE REVIEW

File: components/libc/cplusplus/SConscript [hunk 1]
```diff
@@ -3,10 +3,12 @@
...
 Import('rtconfig')
 
-cwd = GetCurrentDir()
-src = Glob('*.cpp') + Glob('*.c')
+cwd     = GetCurrentDir()
+src     = []
 CPPPATH = [cwd]
 
+if GetDepend('RT_USING_CPP_MULTITHREAD'):
+    src = Glob('*.cpp') + Glob('*.c')
```
Comment (BernardXiong) [inline components/libc/cplusplus/SConscript:11]: 如果没定义 `RT_USING_CPP_MULTITHREAD`，那么基本上这个cplusplus下所有的特性都没了？

File: components/libc/cplusplus/Kconfig [hunk 2]
```diff
@@ -4,6 +4,10 @@ menuconfig RT_USING_CPLUSPLUS
 
 if RT_USING_CPLUSPLUS
 
+    config RT_USING_CPP_MULTITHREAD
+        bool "Enable RT-Thread C++ thread feature support"
```
Comment (mysterywolf) [inline components/libc/cplusplus/Kconfig:8]:
```suggestion
        bool "Enable RT-Thread C++ multi-threads feature support"
```

File: components/libc/cplusplus/SConscript [hunk 3]
```diff
@@ -3,10 +3,14 @@
 from building import *
 Import('rtconfig')
 
-cwd = GetCurrentDir()
-src = Glob('*.cpp') + Glob('*.c')
+cwd     = GetCurrentDir()
+src     = Split('''
+cxx_crt_init.c
+''')
```
Comment (mysterywolf) [inline components/libc/cplusplus/SConscript:9]:
```suggestion
src     = ['cxx_crt_init.c']
```

**Review by mysterywolf — CHANGES_REQUESTED** [review]
https://github.com/RTduino/RTduino/blob/master/core/SConscript#L16

把这一行放进来，可以有效减低开启C++的资源开销

**Comment by mysterywolf** [comment]
我来改吧。

**Review by mysterywolf — APPROVED** [review]
RTduino下编译通过

**Review by Guozhanxin — APPROVED** [review]

**Review by Guozhanxin — APPROVED** [review]

File: bsp/k210/drivers/SConscript [hunk 4]
```diff
@@ -38,7 +38,7 @@ if GetDepend('RT_USING_PWM'):
 if GetDepend('RT_USING_WDT'):
     src += ['drv_wdt.c']
 
-group = DefineGroup('Drivers', src, depend = [''], CPPPATH = CPPPATH)
+group = DefineGroup('Drivers', src, depend = [''], CPPPATH = CPPPATH, CPPDEFINES=['NNCASE_NO_EXCEPTIONS'])
```
Comment (Guozhanxin) [inline bsp/k210/drivers/SConscript:41]: 这个是啥，怎么在这个pr里

Comment (mysterywolf) [reply bsp/k210/drivers/SConscript:41]: 不使用c++ exception 降低资源开销

**Pull request merged**